src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTargetSpelling.java

```diff
@@ -168,6 +182,78 @@ void onDismiss()
          releaseOnDismiss_.remove(0).removeHandler();
    }
 
+   private void injectContextMenuHandler()
+   {
+      docDisplay_.addContextMenuHandler((event) ->
+      {
+         // If we have a selection, just return as the user likely wants to cut/copy/paste
+         if (docDisplay_.hasSelection())
+            return;
+
+         // Get the word under the cursor
+         Position pos = docDisplay_.getCursorPosition();
+         Position endOfLine = Position.create(docDisplay_.getLine(pos.getRow()).length(), pos.getColumn());
+         Iterable<Range> wordSource = docDisplay_.getWords(
+            docDisplay_.getFileType().getTokenPredicate(),
+            docDisplay_.getFileType().getCharPredicate(),
+            pos,
+            endOfLine);
+
+         Iterator<Range> wordsIterator = wordSource.iterator();
+
+         // If there's no word, just return
+         if (!wordsIterator.hasNext())
+            return;
+
+         Range wordRange = wordsIterator.next();
+         String word = docDisplay_.getTextForRange(wordRange);
+
+         if (word.length() < 2 || typoSpellChecker_.checkSpelling(word))
+            return;
+
+         final ToolbarPopupMenu menu = new ToolbarPopupMenu();
+         String[] suggestions = typoSpellChecker_.suggestionList(word);
+
+         // We now know we're going to show our menu, stop default context menu
+         event.preventDefault();
+         event.stopPropagation();
+
+         for (String suggestion : suggestions)
+         {
+            MenuItem suggestionItem = new MenuItem(
+               AppCommand.formatMenuLabel(null, suggestion, ""),
+               true,
+               () -> {
+                  docDisplay_.replaceRange(wordRange, suggestion);
+                  docDisplay_.removeMarkersAtCursorPosition();
+               });
+
+            menu.addItem(suggestionItem);
+         }
+
+         // Only add a separator if we have suggestions to separate from
+         if (suggestions.length > 0)
+            menu.addSeparator();
+
+         MenuItem addToDictionaryItem = new MenuItem(
+            AppCommand.formatMenuLabel(RES.addToDictIcon(), "Add to user dictionary", ""),
+            true,
+            () -> {
+               typoSpellChecker_.addToUserDictionary(word);
+               docDisplay_.removeMarkersAtCursorPosition();
+            });
+
+         menu.addItem(addToDictionaryItem);
+
+
+         menu.setPopupPositionAndShow((offWidth, offHeight) -> {
```

Comment:
Do we need to account for the boundaries of the editor here?  (thinking e.g. source windows where going outside the boundaries of the editor probably means getting clipped)

Reply:
I tried scrolling a lot left/right and it always felt alright to me. I'll play around with it more to confirm.

Reply:
Nvm I didn't account for it cutting off due to being constrained by the chromium window. Got it under control.